Feature/selection

Eve0028 opened the pull request:
1) Selection methods:
- best selection (elite strategy),
- roulette selection,
- tournament selection.
2) Population builder.
3) Tests: Selection methods and population builder.

Eve0028 committed:
doc: Add docstrings and small corrections

Eve0028 committed:
add: Selection methods

Selection methods:
- best selection (elite strategy),
- roulette selection,
- tournament selection

Eve0028 committed:
add: Population builder

Eve0028 committed:
tests: Selection methods and population builder

zannna commented on source/selection/tournament_selection.py:
```diff
@@ -0,0 +1,24 @@
+        """
+        selected_individuals = []
+        for _ in range(selection_size):
+            tournament = random.sample(individuals, tournament_size)
```
to powinno być chyba losowanie bez powtórzeń tych osobników, więc na początku powinno się wybrać wszystkich ziomków i podzielić ich na turnieje

Eve0028 replied:
Ok, racja. Pytanko: Co zrobić jeśli liczba osobników nie dzieli się bez reszty przez parametr tournament_size? Na razie rzucam wyjątkiem

zannna replied:
to znaczy może się nie dzielić tylko selection_size *   tournament_size < liczba osobników 
Jeśli tak nie będzie to wyjątek jest okej

Eve0028 replied:
Wtedy wszyscy osobnicy nie będą wybierani do losowania, i można stracić najlepszych, hmm... Dlatego może wywalę parametr selection_size, i stworzę tyle grup ile się zmieści?

zannna replied:
ja założyłam w krzyżowaniu, że ta lista będzie bez tych najlepszych osobników. Dlatego robię z nią co chce. Wydaje mi się ze to nie będzie cięzkie do ogarniecia w spięciu tego algorytmu

Eve0028 replied:
Ok, przedyskutowane na callu, zamykam

zannna commented on source/selection/roulette_selection.py:
```diff
@@ -0,0 +1,29 @@
+            pick = random.uniform(0, total_fitness)
+            current = 0
+            for individual in individuals:
+                current += individual.count_fitness_function(fitness_function)
```
Jak dla mnie może zostać, ale bardziej po bożemu byłoby gdyby tę dystrybuantę policzyć przed for _ in range(selection_size): Teraz to samo liczenie będzie odbywało się w pętli wielokrotnie. Jeśli selection_size byłoby duże to mogłoby to spowolnić działanie

Eve0028 replied:
Zmienione

Eve0028 committed:
refactor: TournamentSelection - without repetition

Eve0028 committed:
refactor: RouletteSelection - use cumulative distribution

And bisect_left function for binary search

Eve0028 committed:
refactor: TournamentSelection - smaller group

If the last group is smaller than tournament_size, we add it to the previous group.
